### ISSUE TITLE
Failure to restart docker container with DATA_DIR defined and sqs

### ISSUE THREAD

**Issue by shadowspawn** [localstack/localstack#3464]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description

I am running localstack container using `docker-compose` with a volume mounted and `DATA_DIR` defined. I can get the multiple containers up and running from a clean configuration, but encounter problems after restarting. I encountered the problems when using services sqs+dynamodb.

Reproduced with localstack/localstack 0.12.5 and latest. 
macOS Catalina 10.15.7
Docker Engine v20.10.2

Please let me know if there is some additional logging I can supply, or if I should try for a minimal reproduction with a manual copy of a problematic `recorded_api_calls.json` file. 

## Expected behavior

Consistently able to launch services, multiple times.

## Actual behavior

On the second launch, so the persistent data volume has contents, localstack does not come up and displays:

```
Waiting for all LocalStack services to be ready
Waiting for all LocalStack services to be ready
Waiting for all LocalStack services to be ready
...
```

# Steps to reproduce

After the failure occurs (when I also had dynamodb enabled), the problem can be reproduced with just sqs and launching using a simple `docker-compose.yml` file:

```yaml
version: '2.4'
services:
  localstack:
    image: localstack/localstack:latest
    environment:
      - SERVICES=sqs
      - DATA_DIR=/mnt/data
    ports:
      - '4566:4566'
    volumes:
      # Persistent data
      - type: volume
        source: localstack-data
        target: /mnt/data
volumes:
  localstack-data:
```

Launch with `docker-compose up localstack` loops indefinitely waiting for services to be ready.

Connecting to the container and deleting `recorded_api_calls.json` from `DATA_DIR` fixes the problem for the next launch.

## Command used to start LocalStack

```
docker-compose up 
```

**Comment by shadowspawn** [comment]
I am not sure this was happening previously, but today I noticed the services do _eventually_ start.

```
localstack_1             | Ready.
localstack_1             | 2021-01-18T01:55:44:INFO:localstack.utils.persistence: Restored 108 API calls from persistent file: /mnt/data/recorded_api_calls.json
localstack_1             | 2021-01-18T01:55:44:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 365855.30495643616ms
```

**Comment by shadowspawn** [comment]
I am giving up on updating localstack for now, and will close this issue as I am going to pursue other approaches. 

If anyone comes across this soon and has questions, I'm try and answer any questions while I still have the environment available.

**Comment by mgagliardo** [comment]
Hey @shadowspawn , first of all, I was unable to reproduce this. 

Can you please:
-  Give a try to this using `DEBUG=1` as an additional environment variable.
-  On the other hand, `/mnt/data/` does not exist on MacOS (I am using `Catalina 10.15.7` as well) and can't be created as / is a RO filesystem, I am a bit worried it might be either that or permissions on your folder, can you please try something like creating a folder on i.e. `/tmp/localstack`, making sure your current user owns it and pointing `DATA_DIR` there?

Thanks!

**Comment by shadowspawn** [comment]
Some additional information from todays experimentation. On my Mac,  localstack did eventually become ready after 6 min replaying 103 events. For my co-worker on a Windows PC, it took 91 minutes. (Not necessarily the same persistent data.)

**Comment by shadowspawn** [comment]
Thanks @mgagliardo. I'll try `DEBUG=1` tomorrow.

The `/mnt/data/` is the path inside the container I used for the docker managed volume, and does not exist with that path on the Mac host. (See the tail end of the example `docker-compose.yml` file for configuration.)

**Comment by mgagliardo** [comment]
Hi @shadowspawn, yes I missread that, my bad. If you have any findings whatsoever please post them here for further debugging. Cheers.

**Comment by shadowspawn** [comment]
Launching using `docker-compose up localstack` and `DEBUG=1` in environment:

```
localstack_1             | Waiting for all LocalStack services to be ready
localstack_1             | 2021-01-18 20:51:01,864 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
localstack_1             | 2021-01-18 20:51:01,866 INFO supervisord started with pid 14
localstack_1             | 2021-01-18 20:51:02,870 INFO spawned: 'dashboard' with pid 20
localstack_1             | 2021-01-18 20:51:02,873 INFO spawned: 'infra' with pid 21
localstack_1             | 2021-01-18 20:51:02,876 INFO success: dashboard entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
localstack_1             | 2021-01-18 20:51:02,876 INFO exited: dashboard (exit status 0; expected)
localstack_1             | (. .venv/bin/activate; exec bin/localstack start --host)
localstack_1             | Starting local dev environment. CTRL-C to quit.
localstack_1             | LocalStack version: 0.12.4
localstack_1             | 2021-01-18 20:51:04,448 INFO success: infra entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
localstack_1             | Starting edge router (https port 4566)...
localstack_1             | Starting mock DynamoDB service on http port 4566 ...
localstack_1             | Starting mock DynamoDB Streams service on http port 4566 ...
localstack_1             | Starting mock Kinesis service on http port 4566 ...
localstack_1             | Starting mock S3 service on http port 4566 ...
localstack_1             | Listening at http://:::43967
localstack_1             | Initializing DynamoDB Local with the following configuration:
localstack_1             | Port:	38495
localstack_1             | InMemory:	false
localstack_1             | DbPath:	/mnt/data/dynamodb
localstack_1             | SharedDb:	true
localstack_1             | shouldDelayTransientStatuses:	false
localstack_1             | CorsParams:	*
localstack_1             | 
localstack_1             | 2021-01-18T20:51:05:DEBUG:bootstrap.py: Loading plugins - scope "services", module "localstack": <function register_localstack_plugins at 0x7f9c6f7dd550>
localstack_1             | [2021-01-18 20:51:06 +0000] [22] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1             | 2021-01-18T20:51:06:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1             | 2021-01-18T20:51:06:INFO:localstack.multiserver: Starting multi API server process on port 57055
localstack_1             | [2021-01-18 20:51:06 +0000] [22] [INFO] Running on http://0.0.0.0:57055 (CTRL + C to quit)
localstack_1             | 2021-01-18T20:51:06:INFO:hypercorn.error: Running on http://0.0.0.0:57055 (CTRL + C to quit)
localstack_1             | Starting mock SQS service on http port 4566 ...
localstack_1             | Waiting for all LocalStack services to be ready
localstack_1             | Waiting for all LocalStack services to be ready
localstack_1             | Waiting for all LocalStack services to be ready
localstack_1             | Waiting for all LocalStack services to be ready
localstack_1             | Waiting for all LocalStack services to be ready
...
```

**Comment by mgagliardo** [comment]
Have you checked if there's any other process running on your machine that might be using the same port or blocking this? Sometimes cancelling on startup might avoid these services from being teared down. Additionally I would ask you to restart your machine to confirm.

**Comment by shadowspawn** [comment]
The services start fine if I remove DATA_DIR so I don't think it is a port problem, but happy to confirm. No other services running. Anti-virus is disabled. I'll restart and repeat test shortly.

(I'm currently waiting to see if services do eventually start and I can catch the tailing messages with `DEBUG=1`, and I'll add them above.)

**Comment by mgagliardo** [comment]
Can you please try changing `DATA_DIR` to something like `/tmp/localtest`? I have an idea of what it might be.

**Comment by shadowspawn** [comment]
Restarted. Did not notice any change in behaviour, looping waiting for services.

Changed from using docker volume mounted at `/mnt/data` to using local folder `/tmp/localtest` mounted at `/tmp/localtest`. Did not notice any change in behaviour, looping waiting for services. 

Using `docker-compose.yml` file containing:
```
  localstack:
    image: localstack/localstack:0.12.5
    environment:
      - SERVICES=sqs,dynamodb,s3
      - HOSTNAME_EXTERNAL=${MY_DESKTOP_HOST}
      - DATA_DIR=/tmp/localtest
      - DEBUG=1
    ports:
      - '4566:4566'
    volumes:
      - /tmp/localtest:/tmp/localtest
```

**Comment by shadowspawn** [comment]
I did some more experimenting and built-up to a problem state incrementally. The start time blows out after the client application uses the services, which likely involves sending and receiving sqs messages, and adding entries to dyanmodb. 

Erased DATA_DIR. Launched:

```
localstack_1             | 2021-01-18T22:17:23:INFO:localstack.utils.persistence: Restored 1 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-18T22:17:23:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 9229.615926742554ms
```

Shutdown. Launched:

```
localstack_1             | 2021-01-18T22:18:51:INFO:localstack.utils.persistence: Restored 2 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-18T22:18:51:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 9212.355136871338ms
```

Added sqs queues. Shutdown. Launched:

```
localstack_1             | 2021-01-18T22:19:59:INFO:localstack.utils.persistence: Restored 16 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-18T22:19:59:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 9447.039127349854ms
```

Added files to S3. Shutdown. Launched:

```
localstack_1             | 2021-01-18T22:21:57:INFO:localstack.utils.persistence: Restored 212 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-18T22:21:57:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 12257.499933242798ms
```

Ran client application briefly, which uses the localstack services. Shutdown. Launched _slowly_:

```
localstack_1             | 2021-01-18T22:26:55:INFO:localstack.utils.persistence: Restored 237 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-18T22:26:55:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 202972.87917137146ms
```

Casual observation of CPU usage during slow launch:  the CPU usage is around 100% for about 5 s near start of "waiting", then drops down to 1-4% for a long time, then drops to around 0.4% once up.

**Comment by mgagliardo** [comment]
Can you change the `DATA_DIR` and add the `TMPDIR`?  As far as I can see that is the one to be mounted.

```
      - HOST_TMP_FOLDER=${TMPDIR}
    volumes:
      - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
```

**Comment by shadowspawn** [comment]
I did not see any change in behaviour with suggested `HOST_TMP_FOLDER`and `volumes` changes.

Building up to failure again as above, start took 288 s.

```
localstack_1             | 2021-01-19T04:03:21:INFO:localstack.utils.persistence: Restored 243 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1             | 2021-01-19T04:03:21:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 288224.04193878174ms
```

**Comment by shadowspawn** [comment]
I tried running it through docker rather than docker-compose, for another data point. (Mounting one of the volumes prepared earlier rather than building up from empty.)

```
docker run -it -e SERVICES=sqs,dynamodb,s3 -e DATA_DIR=/tmp/localtest --mount type=bind,source=/tmp/localtest,target=/tmp/localtest  localstack/localstack:0.12.5 localstack start
```

```
2021-01-19T04:26:48:INFO:localstack.utils.persistence: Restored 250 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
2021-01-19T04:26:48:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 203311.90490722656ms

```

**Comment by mgagliardo** [comment]
I have tested this out again and unable to reproduce it.

`docker-compose.yml`:
```
version: '2.4'
services:
  localstack:
    image: localstack/localstack:latest
    environment:
      - SERVICES=sqs,dynamodb,s3
      - DATA_DIR=/tmp/localtest
      - DEBUG=1
    ports:
      - '4566:4566'
    volumes:
      - /tmp/localtest:/tmp/localtest
```

DEBUG output:
```
11:58 $ docker-compose up
Creating network "localtest_default" with the default driver
Creating localtest_localstack_1 ... done
Attaching to localtest_localstack_1
localstack_1  | Waiting for all LocalStack services to be ready
localstack_1  | 2021-01-19 10:58:53,741 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
localstack_1  | 2021-01-19 10:58:53,746 INFO supervisord started with pid 15
localstack_1  | 2021-01-19 10:58:54,752 INFO spawned: 'dashboard' with pid 21
localstack_1  | 2021-01-19 10:58:54,758 INFO spawned: 'infra' with pid 22
localstack_1  | 2021-01-19 10:58:54,767 INFO success: dashboard entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
localstack_1  | 2021-01-19 10:58:54,767 INFO exited: dashboard (exit status 0; expected)
localstack_1  | (. .venv/bin/activate; exec bin/localstack start --host)
localstack_1  | 2021-01-19 10:58:55,774 INFO success: infra entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
localstack_1  | Starting local dev environment. CTRL-C to quit.
localstack_1  | LocalStack version: 0.12.5
localstack_1  | Waiting for all LocalStack services to be ready
localstack_1  | Starting edge router (https port 4566)...
localstack_1  | Starting mock DynamoDB service on http port 4566 ...
localstack_1  | 2021-01-19T10:58:59:DEBUG:bootstrap.py: Loading plugins - scope "services", module "localstack": <function register_localstack_plugins at 0x7f476eb61820>
localstack_1  | [2021-01-19 10:59:00 +0000] [23] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1  | 2021-01-19T10:59:00:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1  | Starting mock DynamoDB Streams service on http port 4566 ...
localstack_1  | Starting mock Kinesis service on http port 4566 ...
localstack_1  | Starting mock S3 service on http port 4566 ...
localstack_1  | Initializing DynamoDB Local with the following configuration:
localstack_1  | Port:	34575
localstack_1  | InMemory:	false
localstack_1  | DbPath:	/tmp/localtest/dynamodb
localstack_1  | SharedDb:	true
localstack_1  | shouldDelayTransientStatuses:	false
localstack_1  | CorsParams:	*
localstack_1  | Listening at http://:::50451
localstack_1  |
localstack_1  | 2021-01-19T10:59:00:INFO:localstack.multiserver: Starting multi API server process on port 57831
localstack_1  | [2021-01-19 10:59:00 +0000] [23] [INFO] Running on http://0.0.0.0:57831 (CTRL + C to quit)
localstack_1  | 2021-01-19T10:59:00:INFO:hypercorn.error: Running on http://0.0.0.0:57831 (CTRL + C to quit)
localstack_1  | Starting mock SQS service on http port 4566 ...
localstack_1  | Waiting for all LocalStack services to be ready
localstack_1  | Ready.
localstack_1  | 2021-01-19T10:59:11:INFO:localstack.utils.persistence: Restored 1 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
localstack_1  | 2021-01-19T10:59:11:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 10881.269931793213ms
```

**Comment by mgagliardo** [comment]
I am closing this as I have not been able to reproduce this issue in 3 different computers (2 mac, 1 linux). I would suggest you to:

- update your localstack container to latest
- rm everything on whatever your data_dir is or would be and
- start a clean installation, besides the data_dir issue it is very abnormal the whole API takes +20 times as I did with the same config.

**Comment by shadowspawn** [comment]
Thanks for trying @mgagliardo

Note your reproduction log shows no stored data, and I am not seeing problems in that case either:

```
localstack_1  | 2021-01-19T10:59:11:INFO:localstack.utils.persistence: Restored 1 API calls from persistent file: /tmp/localtest/recorded_api_calls.json
```

I did wonder if the `recorded_api_calls.json` was transferable between computers/containers.

**Comment by mgagliardo** [comment]
I am unsure about it but will try between these computers (and cross OS if possible).

**Comment by marciotoze** [comment]
I have same problem here, I believe that the problem is to recover messages from the volume (maybe is working but just very slow).

the behavior here is simple, we start localstack first time without any problem (via docker) and sent some messages to sqs (in our tests we send about 500 messages). 

now if you down docker conteiner and tries to up again you stay `Waiting for all LocalStack services to be ready` forever, sometimes I just see the `Ready.` message after several minutes (cause that I beliave it just very slow to recover messages).

If I remove docker volume and create a new one, the container just up without any problem again.

Some specs (maybe could be that) im using Windows 10 with WSL2 (ubuntu), Docker is installed on Windows and localstack is started into ubuntu console by a docker-compose file [(as recommended by docker)](https://docs.docker.com/docker-for-windows/wsl/)

**Comment by Kejsty** [comment]
We have the same problem, on Windows with WSL2, Mac OSX, but also on Ubuntu. We use both SQS and SNS quite heavily, and therefore, upon startup, I found this message at the bottom of the log: 

```
Restored 3751 API calls from persistent file: /tmp/localstack/data/recorded_api_calls.json
Execution of "start_api_services" took 189497.98393249512ms
```
Here `recorded_api_calls.json` had ~11Mb, and it took ~3 min.

I assume this is result of following steps:
- v0.11.2: add persistence support to SQS API
- v0.11.3: add persistence support for SNS
- v0.12.4: use locking to fix concurrency during persistence restoration

The last one causes that the localstack API does not respond until startup is done. 

We actually do not benefit from SQS and SNS persistence, we need it for other services such as S3 of dynamo. _Is there some possibility to turn of persistence of SQS and SNS only_?

**Comment by jbarbuto** [comment]
We found this behavior was being caused by not including an AWS account ID in the `Authorization` header. This meant the regexp at https://github.com/localstack/localstack/blob/bc9b10955b6ea254087c013eef244169c8663f80/localstack/utils/aws/aws_stack.py#L188 failed to replace that portion of the header with `INTERNAL_AWS_ACCESS_KEY_ID`, which eventually leads to a deadlock.

If this applies in your case as well, including an account ID (`AKIAIOSFODNN7EXAMPLE` is a commonly used fake value) should resolve the issue.

**Comment by shadowspawn** [comment]
From a slow-start situation, I examined `recorded_api_calls.json`.

All of the lines in `recorded_api_calls.json` have `Credential`, but only 4 of the lines have `__internal_call__` in the line, so I might be in the suggested `INTERNAL_AWS_ACCESS_KEY_ID` failure scenario?

```
/mnt/data # wc <recorded_api_calls.json 
      311     14327   4912836
/mnt/data # grep __internal_call__ recorded_api_calls.json | wc
        4       172      4116
```

The original client calls were generated by the Java AWS SDK, and I do have `AWS_ACCESS_KEY_ID` set in the environment.

**Comment by dmarkey** [comment]
I just stumbled across this issue also using the latest image. Was anyone able to come up with a workaround?

**Comment by shadowspawn** [comment]
I didn't find a work-around. (And have stopped using localstack. Still like the project, but wasn't working for us.)

**Comment by dmarkey** [comment]
Yeah, I feel like this issue should be re-opened. But I guess you'd rather a new issue so you don't get spammed?

**Comment by shadowspawn** [comment]
I can't reopen, and new issue might be better with an original poster still interested in chasing. (But I don't mind the spam if project does wants to reopen this one. Thanks for thinking of that!)

**Comment by alins** [comment]
Is anybody willing to share their `recorded_api_calls.json` file so maintainers can potentially repro this issue?

**Comment by dmarkey** [comment]
I will later on the new issue.

**Comment by newtondev** [comment]
+1 experiencing the same issue trying to start up localstack after stopping it.

**Comment by dmarkey** [comment]
This file worked fine for me locally..

```
{
    "QueueUrls": [
        "http://localhost:4566/000000000000/development-yip-worker-events-low",
        "http://localhost:4566/000000000000/development-yip-worker-events-high",
        "http://localhost:4566/000000000000/community-backend-dead-letter",
        "http://localhost:4566/000000000000/development-community-backend-DeadLetterQueuePolicy-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-DeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-IntegrationDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ConversationDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-QuestionDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ArticleDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-IdeaDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ProductUpdateDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-SubscriptionToMessagebusQueuePolicy-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-CompanyProjectionDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-unified-DeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-yip-worker-events-low-DEAD",
        "http://localhost:4566/000000000000/development-yip-worker-EventsLowDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-yip-worker-events-high-DEAD",
        "http://localhost:4566/000000000000/development-yip-worker-EventsHighDeadLetterQueue-12345ABCDEFGH"
    ]
}
```

**Comment by newtondev** [comment]
Here is another file that has been running for the past 40 minutes:

[api_calls_sqs.json.zip](https://github.com/localstack/localstack/files/6489060/api_calls_sqs.json.zip)

**Comment by newtondev** [comment]
it just completed:

2021-05-16T11:05:10:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 3694539.1278266907ms

This is way too long. Can't wait for this everytime we start the environment.

**Comment by dmarkey** [comment]
Worked OK for me..
```
{
    "QueueUrls": [
        "http://localhost:4566/000000000000/development-yip-worker-events-low",
        "http://localhost:4566/000000000000/development-yip-worker-events-high",
        "http://localhost:4566/000000000000/community-backend-dead-letter",
        "http://localhost:4566/000000000000/development-community-backend-DeadLetterQueuePolicy-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-DeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-IntegrationDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ConversationDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-QuestionDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ArticleDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-IdeaDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-ProductUpdateDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-SubscriptionToMessagebusQueuePolicy-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-legacy-lrp-CompanyProjectionDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-unified-DeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-yip-worker-events-low-DEAD",
        "http://localhost:4566/000000000000/development-yip-worker-EventsLowDeadLetterQueue-12345ABCDEFGH",
        "http://localhost:4566/000000000000/development-yip-worker-events-high-DEAD",
        "http://localhost:4566/000000000000/development-yip-worker-EventsHighDeadLetterQueue-12345ABCDEFGH"
    ]
}
```

```


LocalStack version: 0.12.11

Starting edge router (https port 4566)...
Starting mock S3 service on http port 4566 ...
2021-05-16T12:33:58:INFO:localstack.multiserver: Starting multi API server process on port 61437
[2021-05-16 12:33:58 +0100] [766] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
2021-05-16T12:33:58:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
[2021-05-16 12:33:58 +0100] [766] [INFO] Running on http://0.0.0.0:61437 (CTRL + C to quit)
2021-05-16T12:33:58:INFO:hypercorn.error: Running on http://0.0.0.0:61437 (CTRL + C to quit)
Starting mock SQS service on http port 4566 ...
2021-05-16T12:34:19:INFO:localstack.utils.persistence: Restored 424 API calls from persistent file: /Users/dmarkey/src/localstack_upsteam/data/api_calls_sqs.json
2021-05-16T12:34:19:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 21747.00093269348ms
Ready.
```

**Comment by newtondev** [comment]
Could it have something to do with my environment setup:

macOS Catalina
Docker version 20.10.5, build 55c4c88

```
localstack:
    image: localstack/localstack:latest
    ports:
      - 4566-4599:4566-4599
    environment:
      - HOSTNAME=localstack
      - HOSTNAME_EXTERNAL=localstack
      - DEFAULT_REGION=eu-west-1
      - SERVICES=${LOCALSTACK_SERVICES-iam,lambda,dynamodb,apigateway,s3,sns,sqs}
      - DEBUG=${LOCALSTACK_DEBUG-true}
      - AWS_ACCESS_KEY_ID=test
      - AWS_SECRET_ACCESS_KEY=test
      - INTERNAL_AWS_ACCESS_KEY_ID=${LOCALSTACK_INTERNAL_AWS_ACCESS_KEY_ID-AKIAIOSFODNN7EXAMPLE}
      - DATA_DIR=${LOCALSTACK_DATA_DIR-/tmp/localstack/data}
      - PORT_WEB_UI=${LOCALSTACK_PORT_WEB_UI- }
      - LAMBDA_EXECUTOR=${LOCALSTACK_LAMBDA_EXECUTOR-'docker'}
      - LAMBDA_REMOTE_DOCKER=${LOCALSTACK_LAMBDA_REMOTE_DOCKER-false}
      - LAMBDA_DOCKER_NETWORK=insided
      - HOSTNAME_FROM_LAMBDA=localstack
      - KINESIS_ERROR_PROBABILITY=${LOCALSTACK_KINESIS_ERROR_PROBABILITY- }
      - DOCKER_HOST=unix:///var/run/docker.sock
      - PERSISTENCE_SINGLE_FILE=false
    volumes:
      - localstack:/tmp/localstack
      - /var/run/docker.sock:/var/run/docker.sock:ro
```

**Comment by nri-zhiwei-wang-01** [comment]
Our collegure have the same problem too...
It takes about more than two hours to be Ready status. And the message 'Waiting for all LocalStack services to be ready' repeat 1000+ times.

C:\Users\Administrator>docker logs --tail 5 containers_localstack_1
Waiting for all LocalStack services to be ready
Waiting for all LocalStack services to be ready
Ready.
2021-06-04T07:38:35:INFO:localstack.utils.persistence: Restored 669 API calls from persistent file: /tmp/localstack/data/recorded_api_calls.json
2021-06-04T07:38:35:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 7742829.63514328ms

C:\Users\Administrator>

**Comment by alins** [comment]
any chance to share `recorded_api_calls.json` for repro?

**Comment by Aleksei-Pankratev-EPAM** [comment]
The issue occurs quite stably on my machine:
```
localstack_1      | 2021-08-12 06:42:01,198:API: 127.0.0.1 - - [12/Aug/2021 06:42:01] "POST / HTTP/1.1" 200 -
localstack_1      | 2021-08-12T06:42:01:INFO:localstack.utils.persistence: Restored 32 API calls from persistent file: /tmp/localstack/data/recorded_api_calls.json
localstack_1      | 2021-08-12T06:42:01:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 390778.02ms
localstack_1      | 2021-08-12T06:42:01:DEBUG:localstack.utils.persistence: saving startup info StartupInfo(timestamp='2021-08-12T06:42:01.210242', localstack_version='0.12.16', localstack_ext_version='0.12.14', pro_activated=False)
```
No errors are reported during startup

docker-compose:
```
  localstack:
    image: localstack/localstack:latest
    restart: always
    ports:
      - "12696:4566"
      - "12697:4571"
    environment:
      - DEBUG=1
      - SERVICES=sqs,sns,s3
      - DEFAULT_REGION=us-east-1
      - DATA_DIR=/tmp/localstack/data
    volumes:
      - ./temp/docker_data/localstack:/tmp/localstack
      - ./init_localstack.sh:/docker-entrypoint-initaws.d/init_localstack.sh
```
Docker version 20.10.7, build f0df350
Windows 20H2

[recorded_api_calls.zip](https://github.com/localstack/localstack/files/6973389/recorded_api_calls.zip)

**Comment by zerbitx** [comment]
FWIW I've been experiencing the same issue using localstack in kubernetes and it seems to have been another service that depended on localstack being ready only checking if it got a good response from localstack's /health.  I changed the readinessProbe to

```yaml
          readinessProbe:
            exec:
              command:
                - sh
                - -c
                - >-
                  curl --silent http://localhost/health | grep persistence | grep -q initialized
```
Now (for us) localstack starts up consistently in under 30 seconds.

**Comment by marianobrc** [comment]
@mgagliardo @alins  @dmarkey I'm facing this same issue. I see that a fix has been done in #4033 but looks like it didn't solve this issue for  SQS. I can reproduce it consistently every time. 
```broker_1          | 2022-04-07T20:08:48.602:DEBUG:plugin.manager: instantiating plugin PluginSpec(localstack.aws.provider.sqs:default = <function sqs at 0x7fd2ccca6af0>)
broker_1          | 2022-04-07T20:08:48.602:DEBUG:plugin.manager: loading plugin localstack.aws.provider:sqs:default
broker_1          | 2022-04-07T20:08:48.603:INFO:localstack.services.infra: Starting mock SQS service on http port 4566 ...
broker_1          | 2022-04-07T20:08:48.604:INFO:localstack.services.motoserver: starting moto server on http://0.0.0.0:57741
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Waiting for all LocalStack services to be ready
broker_1          | Ready.
broker_1          | 2022-04-07T20:13:00.292:INFO:localstack.utils.persistence: Restored 68 API calls from persistent file: /tmp/localstack/data/recorded_api_calls.json
broker_1          | 2022-04-07T20:13:00.292:INFO:bootstrap.py: Execution of "preload_services" took 251731.38ms

```

The steps  to reproduce are the same originally described here:

- Start the localstack container and the SQS service using `docker-compose up ` -> The first time runs fast
- Add some messages to the SQS queue
- Restart the localstack container without destroying the volume, for example `docker-compose restart localstack`
- It takes a long time "Waiting for all LocalStack services to be ready" until it starts. The time is proportional to the number of messages added to the queue.

I've tried all the possible solutions suggested here but nothing worked for me.
Is there any known workaround? 
@Aleksei-Pankratev-EPAM have you found a way to disable persistence for SQS maybe?

I can provide any extra info as needed, thnx

**Comment by swhammack** [comment]
For anyone looking for a workaround and using docker-compose, you can try the following:

- In your docker-compose.yml , set the following environment variable `PERSISTENCE_SINGLE_FILE=false`
- Create a script to delete whatever service data you want to delete on startup and then call `/usr/local/bin/docker-entrypoint.sh
`
```
#!/usr/bin/env sh

rm /tmp/localstack/data/api_calls_sqs.json
rm /tmp/localstack/data/api_calls_sns.json

/usr/local/bin/docker-entrypoint.sh
```
- Add script as a volume
- Make that script your entrypoint

Here is an example docker-compose.yml
```
  localstack:
    image: localstack/localstack:0.14.2
    entrypoint: /delete-sns-sqs-data.sh
    environment:
      - SERVICES=s3,sns,sqs
      - PERSISTENCE_SINGLE_FILE=false
      - DATA_DIR=/tmp/localstack/data
      - DOCKER_HOST=unix:///var/run/docker.sock
      - HOST_TMP_FOLDER=${TMPDIR:-/tmp/}localstack
    ports:
      - '6654:4566'
    volumes:
      - ./tmp/localstack:/tmp/localstack
      - /var/run/docker.sock:/var/run/docker.sock
      - ./localstack/init:/docker-entrypoint-initaws.d
      - ./localstack/delete-sns-sqs-data.sh:/delete-sns-sqs-data.sh
```
